src/compiler/checker.ts

```diff
@@ -30513,7 +30519,22 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const jsxOpeningLikeNode = node ;
             const sig = getResolvedSignature(jsxOpeningLikeNode);
             checkDeprecatedSignature(sig, node);
-            checkJsxReturnAssignableToAppropriateBound(getJsxReferenceKind(jsxOpeningLikeNode), getReturnTypeOfSignature(sig), jsxOpeningLikeNode);
+
+            const elementTypeConstraint = getJsxElementTypeTypeAt(jsxOpeningLikeNode);
+            if (elementTypeConstraint !== undefined) {
+                const tagType = isJsxIntrinsicIdentifier(jsxOpeningLikeNode.tagName)
+                    ? // TODO: Should this be a literal that library authors could potentially check against?
+                        stringType

```

Comment:
Constructing a literal here would allow `ElementType = keyof IntrinsicElements  | ...`. Though we already have implicit type-checking of possible intrinsics. Currently, if you pass an unknown intrinsic, TypeScript would already fail with `Property 'my-web-component' does not exist on type 'JSX.IntrinsicElements'.`. That error message is probably already ideal since it points directly at the type you need to adjust: `JSX.IntrinsicElements`.

So it's probably fine to keep the generic string type here?